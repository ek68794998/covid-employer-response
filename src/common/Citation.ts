import { CitationSource } from "./CitationSource";
import { CitationType } from "./CitationType";

export class Citation {
	public positivity: number = 0;

	public sources?: CitationSource[];

	public summary: string = "";

	public type?: CitationType;
}
