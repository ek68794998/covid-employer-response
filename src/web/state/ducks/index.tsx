import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});