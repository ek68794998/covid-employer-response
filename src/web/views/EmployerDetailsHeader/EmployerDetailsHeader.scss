@import "../Colors.scss";
@import "../Dimensions.scss";

$flag-size: 20px;
$icon-size: 32px;
$padding-interior: 12px;

.EmployerDetailsHeader__Icon {
	border-radius: ($icon-size / 2) ($icon-size / 2);
	height: $icon-size;
	margin-right: $padding-interior;
	width: $icon-size;
}

.EmployerDetailsHeader__Location img {
	background: #a1a1a1;
	border-radius: 4px;
	height: ($flag-size * 3 / 4);
	margin-right: 6px;
	width: $flag-size;
}

.EmployerDetailsHeader__Title {
	align-items: center;
	display: flex;
	font-family: 'Baloo Chettan 2';

	h2 {
		font-weight: 400;
		margin: 0 4px;

		a {
			color: inherit;
		}

		+ .EmployerActionLinks__Link {
			margin-left: 12px;
		}
	}

	.EmployerDetailsHeader__TitleGap {
		flex: 1;
	}

	.EmployerDetailsHeader__Rating {
		background: $indicator-neutral;
		border-radius: 16px;
		color: #fff;
		cursor: default;
		font-size: 18px;
		font-weight: 600;
		height: 32px;
		padding: 3px 12px 13px 12px;

		&.EmployerDetailsHeader__Rating--good {
			background: $indicator-positive;
		}

		&.EmployerDetailsHeader__Rating--poor {
			background: $indicator-negative;
		}
	}

	.EmployerDetailsHeader__Ticker {
		color: #1b5e20;
		margin-left: 8px;
		opacity: 0.8;

		&:hover {
			opacity: 1.0;
		}
	}
}

.EmployerDetailsHeader__RatingIcon {
	margin-left: $padding-interior;
	vertical-align: bottom;
}

.EmployerDetailsHeader__Subtitle {
	color: #757575;
	cursor: default;
	display: flex;
	font-size: 12px;
	margin: ($padding-interior / 2) 0 $padding-interior;

	> a, > span {
		align-items: center;
		display: flex;
		margin-right: ($padding-interior * 2);
		transition: 0.2s all;

		&:hover {
			color: #212121;
		}
	}

	a {
		color: inherit;
	}

	i {
		font-size: 18px;
		margin-right: 4px;
		vertical-align: bottom;
	}

	.EmployerDetailsHeader__AggregateRatings {
		align-items: center;
		display: flex;
		flex: 1;
		justify-content: flex-end;
		margin: 0;
		opacity: 0.5;
		transition: 0.2s opacity;
	
		i {
			font-size: 24px;
			margin-right: 2px;
		}
	
		span {
			align-items: center;
			cursor: default;
			display: flex;
			margin-right: 8px;
	
			&.EmployerDetailsHeader__GoodRatings {
				color: $indicator-positive;
			}
	
			&.EmployerDetailsHeader__PoorRatings {
				color: $indicator-negative;
			}
		}
	}
}

@media (max-width: $body-reduced-width) {
	.EmployerDetailsHeader__Title {
		flex-wrap: wrap;
	}

	.EmployerDetailsHeader__Title--noShort {
		.EmployerDetailsHeader__Subtitle {
			> a, > span {
				display: block;
				margin-top: 8px;
			}
		}

		.EmployerDetailsHeader__Links i {
			margin: 8px 6px 0 0 !important;
		}
	
		.EmployerDetailsHeader__LinkBreak {
			flex-basis: 100%;
			height: 0;
		}
	}

	.EmployerDetailsHeader__RatingIcon {
		display: none;
	}
}

@media (any-hover: none) {
	.EmployerDetailsHeader__Subtitle .EmployerDetailsHeader__AggregateRatings {
		opacity: 1.0 !important;
	}
}
