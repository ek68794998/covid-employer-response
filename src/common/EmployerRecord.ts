import { Citation } from "./Citation";
import { EmployerLocation } from "./EmployerLocation";
import { EmployerRating } from "./EmployerRating";

export class EmployerRecord {
	public citations: Citation[] = [];

	public employeesBeforeMax: number = -1;

	public employeesBeforeMin: number = -1;

	public id: string = "";

	public image?: string;

	public location: EmployerLocation = new EmployerLocation();

	public name: string = "";

	public officialWebsite?: string;

	public rating: EmployerRating = "fair";

	public summary: string = "";

	public wiki?: string;
}
