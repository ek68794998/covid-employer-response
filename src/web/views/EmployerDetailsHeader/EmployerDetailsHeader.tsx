import React from "react";
import { useSelector } from "react-redux";
import { RouteProps } from "react-router-dom";

import { DesignHelpers } from "../../../common/DesignHelpers";
import { EmployerEmployeeProfile } from "../../../common/EmployerEmployeeProfile";
import { EmployerLocation } from "../../../common/EmployerLocation";
import { EmployerRecordBase } from "../../../common/EmployerRecordBase";
import { EmployerRecordMetadata } from "../../../common/EmployerRecordMetadata";
import { format, LocalizedStrings } from "../../../common/LocalizedStrings";
import { WikipediaHelpers } from "../../../common/WikipediaHelpers";

import { getStrings } from "../../state/ducks/localization/selectors";

import EmployerActionLinks from "../EmployerActionLinks/EmployerActionLinks";
import EmployerLogo from "../EmployerLogo/EmployerLogo";

import "./EmployerDetailsHeader.scss";

interface Props extends RouteProps {
	employer: EmployerRecordMetadata;

	onClickEmployerName?: () => void;

	useShortText?: boolean;
}

const getAliasesComponent =
	(employer: EmployerRecordMetadata, strings: LocalizedStrings, useShortText: boolean): JSX.Element | null => {
		if (!employer.aliases || employer.aliases.length === 0 || useShortText) {
			return null;
		}

		let text: JSX.Element;

		const separator: string = ", ";
		const cutoff: number = 3;

		if (employer.aliases.length > cutoff) {
			const moreText: string = format(strings.more, { number: employer.aliases.length - cutoff });

			text = (
				<>
					{employer.aliases.slice(0, cutoff).join(separator)}&hellip;
					<span
						className="EmployerDetailsHeader__AliasesMore"
						title={employer.aliases.slice(cutoff).join(separator)}
					>
						({moreText})
					</span>
				</>
			);
		} else {
			text = <>{employer.aliases.join(separator)}</>;
		}

		return (
			<span title={strings.detailDescriptions.aka}>
				{DesignHelpers.materialIcon("alternate_email")}
				{text}
			</span>
		);
	};

const getEmployeeCountComponent =
	(employer: EmployerRecordMetadata, strings: LocalizedStrings, useShortText: boolean): JSX.Element | null => {
		const employeeCountString: string | null =
			employer.employeesBefore
				? EmployerEmployeeProfile.toString(employer.employeesBefore, !useShortText, useShortText)
				: null;

		if (!employeeCountString) {
			return null;
		}

		return (
			<span title={strings.detailDescriptions.employees}>
				{DesignHelpers.materialIcon("people")}
				{employeeCountString}
			</span>
		);
	};

const getLocationWikipediaComponent =
	(employer: EmployerRecordMetadata, strings: LocalizedStrings, useShortText: boolean): JSX.Element | null => {
		if (!employer.location) {
			return null;
		}

		const locationString: string =
			useShortText
				? employer.location.city
				: EmployerLocation.toString(employer.location, strings.countryNames);

		const locationWikipediaUrl: string | null = WikipediaHelpers.getWikipediaUrl(employer.location.wiki);

		if (!locationWikipediaUrl) {
			return <span>{locationString}</span>;
		}

		return (
			<a
				className="EmployerDetailsHeader__Location"
				href={locationWikipediaUrl}
				rel="noopener noreferrer"
				target="_blank"
				title={strings.detailDescriptions.location}
			>
				<img
					src={`/images/flags/${employer.location.country}.svg`}
					title={strings.countryNames[employer.location.country]}
				/>
				{locationString}
			</a>
		);
	};

const getTickerComponent =
	(employer: EmployerRecordMetadata, strings: LocalizedStrings): JSX.Element | null => {
		if (!employer.ticker) {
			return null;
		}

		const tickerUrl: string = `https://finance.yahoo.com/quote/${employer.ticker}`;

		return (
			<a
				className="EmployerDetailsHeader__Ticker"
				href={tickerUrl}
				rel="noopener noreferrer"
				target="_blank"
				title={strings.detailDescriptions.ticker}
			>
				(${employer.ticker})
			</a>
		);
	};

const EmployerDetailsHeader: React.FC<Props> = (props: Props): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);
	const { employer, onClickEmployerName, useShortText } = props;

	const displayName: string = useShortText && employer.shortName || employer.name;

	const employerNameComponent: JSX.Element =
		onClickEmployerName
			? <a onClick={onClickEmployerName} title={employer.name !== displayName ? employer.name : ""}>{displayName}</a>
			: <>{displayName}</>;

	const logoImageRegex: RegExpExecArray | null =
		employer.image ? EmployerRecordBase.IMAGE_REGEX.exec(employer.image) : null;

	return (
		<>
			<div className={`EmployerDetailsHeader__Title ${useShortText ? "" : "EmployerDetailsHeader__Title--noShort"}`}>
				<EmployerLogo employer={employer} />
				<h2>
					{employerNameComponent}
					{!useShortText && getTickerComponent(employer, strings)}
				</h2>
				{!useShortText && <EmployerActionLinks employer={employer} />}
				<span className="EmployerDetailsHeader__TitleGap" />
				<span
					className={`EmployerDetailsHeader__Rating EmployerDetailsHeader__Rating--${employer.rating}`}
					title={strings.detailDescriptions.rating}
				>
					{strings.ratingLabels[employer.rating]}
					{DesignHelpers.materialIcon(
						EmployerRecordMetadata.getTrendIcon(employer), "EmployerDetailsHeader__RatingIcon")}
				</span>
			</div>
			<div className="EmployerDetailsHeader__Subtitle">
				{getLocationWikipediaComponent(employer, strings, useShortText || false)}
				{getAliasesComponent(employer, strings, useShortText || false)}
				{getEmployeeCountComponent(employer, strings, useShortText || false)}
				<span className="EmployerDetailsHeader__AggregateRatings" title={strings.detailDescriptions.ratingCounts}>
					<span className="EmployerDetailsHeader__GoodRatings">
						<i className="material-icons">add</i>
						{employer.positiveCount}
					</span>
					<span className="EmployerDetailsHeader__PoorRatings">
						<i className="material-icons">remove</i>
						{employer.negativeCount}
					</span>
				</span>
			</div>
		</>
	);
};

export default EmployerDetailsHeader;
