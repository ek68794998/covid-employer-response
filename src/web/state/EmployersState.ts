import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}