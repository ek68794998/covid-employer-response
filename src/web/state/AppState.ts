import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}