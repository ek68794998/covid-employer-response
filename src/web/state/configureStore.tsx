import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;