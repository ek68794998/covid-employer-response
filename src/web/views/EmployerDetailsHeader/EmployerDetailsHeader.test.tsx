import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { mockComponent, ploc } from "../../../__tests__/TestUtils";
import { EmployerRecord } from "../../../common/EmployerRecord";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import EmployerDetailsHeader from "./EmployerDetailsHeader";

jest.mock(
	"../EmployerActionLinks/EmployerActionLinks",
	() => mockComponent("EmployerActionLinks"));

describe("<EmployerDetailsHeader />", () => {
	test("renders without exploding", () => {
		const store: Store<AppState, AnyAction> = configureStore({
			strings: {
				detailDescriptions: {
					aka: ploc("aka"),
					employees: ploc("employees"),
					location: ploc("location"),
					rating: ploc("rating"),
					ratingCounts: ploc("ratingCounts"),
					ticker: ploc("ticker"),
				},
				ratingLabels: {
					fair: ploc("fair"),
					good: ploc("good"),
					poor: ploc("poor"),
				},
			},
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerDetailsHeader
							employer={new EmployerRecord()}
							onClickEmployerName={(): void => { /* Do nothing. */ }}
						/>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
