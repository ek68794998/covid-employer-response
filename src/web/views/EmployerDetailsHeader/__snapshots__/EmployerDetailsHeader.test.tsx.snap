// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerDetailsHeader /> renders without exploding 1`] = `
Array [
  <div
    className="EmployerDetailsHeader__Title EmployerDetailsHeader__Title--noShort"
  >
    <h2>
      <a
        href="#"
        onClick={[Function]}
        title=""
      >
        
      </a>
    </h2>
    <span
      className="EmployerDetailsHeader__LinkBreak"
    />
    <span
      className="EmployerDetailsHeader__Links"
    />
    <span
      className="EmployerDetailsHeader__Rating EmployerDetailsHeader__Rating--fair"
      title="rating string"
    >
      fair string
      <i
        className="material-icons EmployerDetailsHeader__RatingIcon"
      >
        trending_flat
      </i>
    </span>
  </div>,
  <div
    className="EmployerDetailsHeader__Subtitle"
  />,
]
`;
